### DISCUSSION TITLE
Add utilities to flip an element

### DISCUSSION THREAD

**Discussion by pryley** [tailwindlabs/tailwindcss#2146] [Ideas]
For example:

```css
/* Flip on the X axis */
.flip-horizontal {
    --transform-scale-x: -1;
}
```

```css
/* Flip on the Y axis */
.flip-vertical {
    --transform-scale-y: -1;
}
```

```html
<div class="flip-vertical transform">This is flipped vertically</div>
```

**Comment by adamwathan** [comment]
You can actually do this already by adding `-1` to your scale scale (heh) in your config file:

```js
module.exports = {
  theme: {
    extend: {
      scale: {
        '-1': '-1'
      }
    }
  }
}
```

**Reply by basepack** [reply]
Unfortunately this would not flip SVG's. It produces this code:
```
--tw-scale-x: -1;
--tw-scale-y: -1;
```
Where we need a `transform: scale(-1,1)` to get a SVG flipped horizontally.

**Comment by pryley** [comment]
I've been doing this (the utility names are more verbose):

```js
plugin(function({ addUtilities, theme, variants }) {
  const newUtilities = {
    '.flip-horizontal': {
      '--transform-scale-x': '-1',
    },
    '.flip-vertical': {
      '--transform-scale-y': '-1',
    },
  }
  addUtilities(newUtilities, variants('flip'))
})
```

**Comment by pryley** [comment]
You can close this if you feel it's unnecessary, but I think it would be nice to have explicitly-named utilities for flipping elements; `--transform-scale-{AXIS}: -1` isn't immediately obvious what it does unless you already know.

**Comment by jesperp** [comment]
I was a bit surprised you couldn't flip using Tailwind, not even through overriding! Overriding scale wont work, it flips both x and y axis!

So a `flip-x` and `flip-y` would be awesome!

**Comment by pryley** [comment]
@jesperp Using the example @adamwathan provided above, you should be able to do this: `-scale-x-1` and `-scale-y-1`

**Reply by jesperp** [reply]
Ah, didn't see that, makes complete sense! Thanks

**Comment by SjorsO** [comment]
With JIT, you can use `scale-x-[-1]` to flip an element horizontally

**Reply by fabswt** [reply]
Or just `-scale-x-100`.

**Comment by windyclory** [comment]
My solution:
```
scale: {
  flip: '-1',
},
```
Classes: `scale-x-flip` and `scale-y-flip`

**Reply by leo** [reply]
This works great. Thank you!

**Comment by alikasmou** [comment]
-scale-y-100 -rotate-180

**Reply by Maybach91** [reply]
For `svg` this is not working for me. The following worked for me tho for `svg` in Tailwind `3.4.17`:
`-scale-x-100 !scale-y-100`

**Comment by erlandsona** [comment]
Any thoughts on how to translate this to v4?

**Reply by rozsazoltan** [reply]
Yeah, just use `@utility` directive:
* https://play.tailwindcss.com/9WTF5WWAPg?file=css (I do not recommend it)

```css
@utility flip-horizontal {
  scale: -100% 1;
}
@utility flip-vertical {
  scale: 1 -100%;
}
```

```html
<div class="transform flip-horizontal">
  Hello World
</div>

<div class="transform flip-vertical">
  Hello World
</div>
```

But I wouldn't create a new `flip-*` utility just for this, because IntelliSense won't be able to associate it with `scale-*`. As a result, using `flip-horizontal` together with `scale-x-50` is a logical mistake, and you won't get any warning about it. I would use scale with -100 values instead:
* https://play.tailwindcss.com/SkCaEBV2JX (**recommended**)

```html
<div class="transform -scale-x-100">
  Hello World Horizontal
</div>

<div class="transform -scale-y-100">
  Hello World Vertical
</div>
```

However, you can extend `scale-*` with custom names, so you can add horizontal and vertical as well. Unfortunately, in the case of `scale`, you need the `-x`, `-y`, or `-z` prefix so it knows which axis you want the custom name to apply to. Because of that, I would only define the `-100%` value under a custom name like full, so `scale-x-full` would be the horizontal flip, and `scale-y-full` would be the vertical one.
* https://play.tailwindcss.com/udhDj2HaBg?file=css (**recommended**)

```css
@theme {
  --scale-full: -100%;
}
```

```html
<div class="transform scale-x-full">
  Hello World Horizontal
</div>

<div class="transform scale-y-full">
  Hello World Vertical
</div>
```

Or you can generate names without the `-x` and `-y` parts by duplicating code, but unfortunately this requires fully duplicating the entire scale configuration.
* https://play.tailwindcss.com/GC2OpfhBUt?file=css

```css
@utility scale-horizontal {
  scale: -100% 1;
}
@utility scale-vertical {
  scale: 1 -100%;
}
```

```html
<div class="transform scale-horizontal">
  Hello World
</div>

<div class="transform scale-vertical">
  Hello World
</div>
```

**Comment by ghost** [comment]
How do I unsubscribe?

**Reply by Maybach91** [reply]
You will find the button on the discussion page on the top. You have to visit the discussion page tho. @michael-stout-ai
